### PR TITLE
 authority validation update cache

### PR DESCRIPTION
**Pull request by weijjia**
Please make sure every Pull Request has the following information:

<ul>
  <li>#913</li>
  <li>Cache update for authority migration. New token will always be written into the preferred cache location; For read, try preferred_cache first, followed by provided authority and aliased authority. </li>
</ul>

### CODE REVIEW

File: adal/src/androidTest/java/com/microsoft/aad/adal/DiscoveryTests.java [hunk 1]
```diff
@@ -311,6 +311,51 @@ public Void call() throws Exception {
...
 
+    /**
+     * Verified scenario:
+     * When an authority is valid and metadata is returned:
+     * a. Subsequent requests for any aliases provided in the metadata do not result in further validation network requests
+     * being made for the process lifetime
+     */
+    @Test
+    public void testAuthorityInAliasedList() throws IOException {
+        AuthorityValidationMetadataCache.clearAuthorityValidationCache();
```
Comment (iambmelt) [inline adal/src/androidTest/java/com/microsoft/aad/adal/DiscoveryTests.java]: Nit: Seems more appropriate to place in a `setUp()` function

File: adal/src/main/java/com/microsoft/aad/adal/Utility.java [hunk 2]
```diff
@@ -46,4 +50,10 @@ static boolean isClaimsChallengePresent(final AuthenticationRequest request) {
         // if developer pass claims down through extra qp, we should also skip cache.
         return !StringExtensions.isNullOrBlank(request.getClaimsChallenge());
     }
+
+    static URL constructAuthorityUrl(final URL originalAuthority, final String host) throws MalformedURLException {
+        final String path = originalAuthority.getPath().replaceFirst("/", "");
+        final Uri.Builder builder = new Uri.Builder().scheme(originalAuthority.getProtocol()).authority(host).appendPath(path);
+        return new URL(builder.build().toString());
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/Utility.java]: nit: you can just do builder.build().toURL

Comment (addev-ashish) [reply adal/src/main/java/com/microsoft/aad/adal/Utility.java]: another nit: there was some thing about commenting static routines.

Comment (weijjia) [reply adal/src/main/java/com/microsoft/aad/adal/Utility.java]: can you elaborate a bit more on commenting static routines?

File: adal/src/androidTest/java/com/microsoft/aad/adal/DiscoveryTests.java (hunk 1 above)
Comment (weijjia) [reply adal/src/androidTest/java/com/microsoft/aad/adal/DiscoveryTests.java]: done.

File: adal/src/main/java/com/microsoft/aad/adal/Utility.java (hunk 2 above)
Comment (weijjia) [reply adal/src/main/java/com/microsoft/aad/adal/Utility.java]: I don't think i can do a toURL with Uri.builder.build

**Pull request merged**